### PR TITLE
feat(test): use ort 1.14

### PR DESCRIPTION
**Pull request by justinchuby**
Use ort 1.14 in test

Fix errors in logsumexp and upsample_nearest2d and enable their tests
Parameterized and refactored the backend tests.

### CODE REVIEW

**Review by titaiwangms — APPROVED** [review]

**Review by titaiwangms — COMMENTED** [review]
CI seems far from ready.

**Comment by justinchuby** [comment]
@gramalingam @abock @xadupre _I think_ we are seeing some segfaults in ort (https://github.com/microsoft/onnx-script/actions/runs/4169248601/jobs/7216998170#step:5:976) and new errors when running OptionalGetElement with opset18.

**Comment by justinchuby** [comment]
```
onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py::TestOutputConsistencyCPU::test_output_match_logsumexp_cpu_float32 Fatal Python error: Aborted

Current thread 0x00007f11c7772740 (most recent call first):
  File "/home/justinchu/anaconda3/envs/onnx/lib/python3.10/site-packages/onnxruntime/capi/onnxruntime_inference_collection.py", line 200 in run
  File "/home/justinchu/dev/onnx-script/onnxscript/evaluator.py", line 384 in _call_ort
  File "/home/justinchu/dev/onnx-script/onnxscript/evaluator.py", line 407 in _eval
  File "/home/justinchu/dev/onnx-script/onnxscript/evaluator.py", line 145 in eval
  File "/home/justinchu/dev/onnx-script/onnxscript/values.py", line 185 in __call__
  File "/home/justinchu/dev/onnx-script/onnxscript/onnx_opset/_impl/opset18.py", line 1128 in ReduceLogSumExp
  File "/home/justinchu/dev/onnx-script/onnxscript/function_libs/torch_aten/ops/core.py", line 3053 in aten_logsumexp
  File "/home/justinchu/dev/onnx-script/onnxscript/evaluator.py", line 235 in eval_function
  File "/home/justinchu/dev/onnx-script/onnxscript/values.py", line 295 in __call__
  File "/home/justinchu/dev/onnx-script/onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py", line 695 in test_output_match
  File "/home/justinchu/anaconda3/envs/onnx/lib/python3.10/site-packages/torch/testing/_internal/common_device_type.py", line 815 in test_wrapper
  File "/home/justinchu/anaconda3/envs/onnx/lib/python3.10/site-packages/torch/testing/_internal/common_device_type.py", line 378 in instantiated_test
  File "/home/justinchu/anaconda3/envs/onnx/lib/python3.10/unittest/case.py", line 549 in _callTestMethod
  File "/home/justinchu/anaconda3/envs/onnx/lib/python3.10/unittest/case.py", line 591 in run
  File "/home/justinchu/anaconda3/envs/onnx/lib/python3.10/site-packages/torch/testing/_internal/common_utils.py", line 2038 in _run_with_retry
  File "/home/justinchu/anaconda3/envs/onnx/lib/python3.10/site-packages/torch/testing/_internal/common_utils.py", line 2096 in run
  File "/home/justinchu/anaconda3/envs/onnx/lib/python3.10/site-packages/torch/testing/_internal/common_device_type.py", line 438 in run
  File "/home/justinchu/anaconda3/envs/onnx/lib/python3.10/unittest/case.py", line 650 in __call__
  File "/home/justinchu/anaconda3/envs/onnx/lib/python3.10/site-packages/_pytest/unittest.py", line 330 in runtest
  File "/home/justinchu/anaconda3/envs/onnx/lib/python3.10/site-packages/_pytest/runner.py", line 167 in pytest_runtest_call
  File "/home/justinchu/anaconda3/envs/onnx/lib/python3.10/site-packages/pluggy/_callers.py", line 39 in _multicall
  File "/home/justinchu/anaconda3/envs/onnx/lib/python3.10/site-packages/pluggy/_manager.py", line 80 in _hookexec
  File "/home/justinchu/anaconda3/envs/onnx/lib/python3.10/site-packages/pluggy/_hooks.py", line 265 in __call__
  File "/home/justinchu/anaconda3/envs/onnx/lib/python3.10/site-packages/_pytest/runner.py", line 260 in <lambda>
  File "/home/justinchu/anaconda3/envs/onnx/lib/python3.10/site-packages/_pytest/runner.py", line 339 in from_call
  File "/home/justinchu/anaconda3/envs/onnx/lib/python3.10/site-packages/_pytest/runner.py", line 259 in call_runtest_hook
  File "/home/justinchu/anaconda3/envs/onnx/lib/python3.10/site-packages/_pytest/runner.py", line 220 in call_and_report
  File "/home/justinchu/anaconda3/envs/onnx/lib/python3.10/site-packages/_pytest/runner.py", line 131 in runtestprotocol
  File "/home/justinchu/anaconda3/envs/onnx/lib/python3.10/site-packages/_pytest/runner.py", line 112 in pytest_runtest_protocol
  File "/home/justinchu/anaconda3/envs/onnx/lib/python3.10/site-packages/pluggy/_callers.py", line 39 in _multicall
  File "/home/justinchu/anaconda3/envs/onnx/lib/python3.10/site-packages/pluggy/_manager.py", line 80 in _hookexec
  File "/home/justinchu/anaconda3/envs/onnx/lib/python3.10/site-packages/pluggy/_hooks.py", line 265 in __call__
  File "/home/justinchu/anaconda3/envs/onnx/lib/python3.10/site-packages/_pytest/main.py", line 349 in pytest_runtestloop
  File "/home/justinchu/anaconda3/envs/onnx/lib/python3.10/site-packages/pluggy/_callers.py", line 39 in _multicall
  File "/home/justinchu/anaconda3/envs/onnx/lib/python3.10/site-packages/pluggy/_manager.py", line 80 in _hookexec
  File "/home/justinchu/anaconda3/envs/onnx/lib/python3.10/site-packages/pluggy/_hooks.py", line 265 in __call__
  File "/home/justinchu/anaconda3/envs/onnx/lib/python3.10/site-packages/_pytest/main.py", line 324 in _main
  File "/home/justinchu/anaconda3/envs/onnx/lib/python3.10/site-packages/_pytest/main.py", line 270 in wrap_session
  File "/home/justinchu/anaconda3/envs/onnx/lib/python3.10/site-packages/_pytest/main.py", line 317 in pytest_cmdline_main
  File "/home/justinchu/anaconda3/envs/onnx/lib/python3.10/site-packages/pluggy/_callers.py", line 39 in _multicall
  File "/home/justinchu/anaconda3/envs/onnx/lib/python3.10/site-packages/pluggy/_manager.py", line 80 in _hookexec
  File "/home/justinchu/anaconda3/envs/onnx/lib/python3.10/site-packages/pluggy/_hooks.py", line 265 in __call__
  File "/home/justinchu/anaconda3/envs/onnx/lib/python3.10/site-packages/_pytest/config/__init__.py", line 167 in main
  File "/home/justinchu/anaconda3/envs/onnx/lib/python3.10/site-packages/_pytest/config/__init__.py", line 190 in console_main
  File "/home/justinchu/anaconda3/envs/onnx/bin/pytest", line 8 in <module>
```

**Comment by justinchuby** [comment]
`test_onnxfns_space_to_depth` seems to be failing sporadically with SegFaults: https://github.com/microsoft/onnx-script/actions/runs/4177117953/jobs/7234296466#step:5:1008. I am not feeling super good about this

**Comment by justinchuby** [comment]
Happens when running pytest in parallel

**Comment by justinchuby** [comment]
https://github.com/onnx/onnx/pull/4905

**Comment by justinchuby** [comment]
Thankfully it's not optional has element which we definitely want to use. @gramalingam is it ok if we skipped this test? What would be the appropriate way to do it?

**Comment by gramalingam** [comment]
> Thankfully it's not optional has element which we definitely want to use. @gramalingam is it ok if we skipped this test? What would be the appropriate way to do it?

You mean SpaceToDepth test? Yes, please feel free to skip that test. I guess we could use a `unittest.skipIf(True, ...)` ? Seems slightly preferable to commenting it out. 

Anyway, we need to figure out a good answer to the general question of using conditional statements where one branch may fail certain onnx checks (because it represents a branch that won't execute) for the long term (before we can fully handle such test-cases).

File: onnxscript/tests/functions/onnxfns2_test.py [hunk 1]
```diff
@@ -115,6 +115,10 @@ def test_onnxfns_hardmax(self):
     #         mode=default_mode,
     #         skip_test_names=[])
 
+    @unittest.skipIf(
+        Version(onnxruntime.__version__) == Version("1.14"),
```
Comment (gramalingam) [inline onnxscript/tests/functions/onnxfns2_test.py:119]: May be `<=` would be better than `==`

Comment (justinchuby) [reply onnxscript/tests/functions/onnxfns2_test.py:119]: It was successful in 1.13

Comment (justinchuby) [reply onnxscript/tests/functions/onnxfns2_test.py:119]: Hmm I think this is another issue that is not OptionalGetElement.

**Comment by justinchuby** [comment]
@gramalingam OptionalGetElement is a failing backend test. Is there a good way to disable it?

```
onnxscript/tests/onnx_backend_test.py::TestOnnxBackEnd::test_enumerate_onnx_tests_run - RuntimeError: Unable to create onnxruntime InferenceSession for executing .OptionalGetElement op with onnx model
<
   ir_version: 8,
   opset_import: ["" : 18]
>
node_graph (float[4] input0) => ( output0) {
   output0 = OptionalGetElement (input0)
}
```

**Comment by justinchuby** [comment]
So the segfault was not OptionalGetElement.

OptionalGetElement causes

```
E   onnxruntime.capi.onnxruntime_pybind11_state.Fail: [ONNXRuntimeError] : 1 : FAIL : Node () output arg (output0) type inference failed
```

The segfault in test_onnxfns_space_to_depth does not have OptionalGetElement.

File: onnxscript/tests/functions/onnxfns2_test.py (hunk 1 above)
Comment (gramalingam) [reply onnxscript/tests/functions/onnxfns2_test.py:119]: Oh, I see! I guess we need to investigate it then. (Agree the condition above seems appropriate in this case.)

**Comment by justinchuby** [comment]
@xadupre I updated the backend tests so they are easier to skip. Please take a look.

File: onnxscript/function_libs/torch_aten/ops/core.py [hunk 2]
```diff
@@ -3072,7 +3072,12 @@ def aten_logspace(start: float, end: float, steps: int, base: float = 10.0) -> T
 def aten_logsumexp(self: TReal, dim: INT64, keepdim: int = False) -> TReal:
     """logsumexp(Tensor self, int[1] dim, bool keepdim=False) -> Tensor"""
 
-    return op.ReduceLogSumExp(self, dim, keepdims=keepdim)
+    if op.Size(op.Shape(self)) == 0:
+        # A scalar
+        result = op.Identity(self)
```
Comment (fatcat-z) [inline onnxscript/function_libs/torch_aten/ops/core.py:3077]: Why do we need to add op.Identity here? What's the different between it and returning self directly?

Comment (justinchuby) [reply onnxscript/function_libs/torch_aten/ops/core.py]: removed

Comment (justinchuby) [reply onnxscript/function_libs/torch_aten/ops/core.py]: they should be the same

**Pull request merged**